src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,567 @@
+        mov             rdx,  [rsp + TheUMEntryPrestub_STACK_FRAME_SIZE + 18h]
+        mov             rcx,  [rsp + TheUMEntryPrestub_STACK_FRAME_SIZE + 20h]
+
+        movdqa          xmm0, [rsp + TheUMEntryPrestub_XMM_SAVE_OFFSET]
+        movdqa          xmm1, [rsp + TheUMEntryPrestub_XMM_SAVE_OFFSET + 10h]
+        movdqa          xmm2, [rsp + TheUMEntryPrestub_XMM_SAVE_OFFSET + 20h]
+        movdqa          xmm3, [rsp + TheUMEntryPrestub_XMM_SAVE_OFFSET + 30h]
+
+        //
+        // epilogue
+        //
+        add             rsp, TheUMEntryPrestub_STACK_FRAME_SIZE
+        TAILJMP_RAX
+        
+NESTED_END TheUMEntryPrestub, _TEXT
+
+
+//
+// METHODDESC_REGISTER: UMEntryThunk*
+//
+NESTED_ENTRY UMThunkStub, _TEXT, UMThunkStubUnwindFrameChainHandler
+
+// number of integer registers saved in prologue
+#define UMThunkStubAMD64_NUM_REG_PUSHES 2
+
+// rare path spill area
+#define UMThunkStubAMD64_RARE_PATH_SPILL_SIZE 0x10
+#define UMThunkStubAMD64_RARE_PATH_SPILL_NEGOFFSET 0x20
+
+// HOST_NOTIFY_FLAG
+#define UMThunkStubAMD64_HOST_NOTIFY_FLAG_NEGOFFSET 0x28
+
+// XMM save area 
+#define UMThunkStubAMD64_XMM_SAVE_NEGOFFSET 0x68
+
+// Add in the callee scratch area size.
+#define UMThunkStubAMD64_CALLEE_SCRATCH_SIZE SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES
+#define UMThunkStubAMD64_STACK_FRAME_SIZE 0x88
+
+// Now we have the full size of the stack frame.  The offsets have been computed relative to the
+// top, so negate them to make them relative to the post-prologue rsp.
+#define UMThunkStubAMD64_FRAME_OFFSET UMThunkStubAMD64_CALLEE_SCRATCH_SIZE
+#define UMThunkStubAMD64_RARE_PATH_SPILL_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE - UMThunkStubAMD64_FRAME_OFFSET - UMThunkStubAMD64_RARE_PATH_SPILL_NEGOFFSET)
+#define UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE - UMThunkStubAMD64_FRAME_OFFSET - UMThunkStubAMD64_HOST_NOTIFY_FLAG_NEGOFFSET)
+#define UMThunkStubAMD64_XMM_SAVE_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE - UMThunkStubAMD64_FRAME_OFFSET - UMThunkStubAMD64_XMM_SAVE_NEGOFFSET)
+#define UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE + 8 - UMThunkStubAMD64_FRAME_OFFSET)
+#define UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE (UMThunkStubAMD64_STACK_FRAME_SIZE - (UMThunkStubAMD64_NUM_REG_PUSHES * 8))
+
+#if (UMTHUNKSTUB_HOST_NOTIFY_FLAG_RBPOFFSET - UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET) != 0
+#error update UMTHUNKSTUB_HOST_NOTIFY_FLAG_RBPOFFSET
+#endif
+
+//
+// [ callee scratch ]            <-- new RSP
+// [ callee scratch ]
+// [ callee scratch ]
+// [ callee scratch ]
+// {optional stack args passed to callee}
+// xmm0                          <-- RBP
+// xmm1
+// xmm2
+// xmm3
+// {optional padding to align xmm regs}
+// HOST_NOTIFY_FLAG (needs to make ReverseLeaveRuntime call flag)
+// [rare path spill area]
+// [rare path spill area]
+// rbp save
+// r12 save
+// return address                <-- entry RSP
+// [rcx home]
```

Comment:
As far as I remember, there is no [rcx home] according to the System V specs. Same applies to rdx, r8, r9.